src/api/artifact/controller.go

```diff
@@ -507,6 +512,48 @@ func (c *controller) RemoveLabel(ctx context.Context, artifactID int64, labelID
 	return c.labelMgr.RemoveFrom(ctx, labelID, artifactID)
 }
 
+func (c *controller) Walk(ctx context.Context, root *Artifact, walkFn func(*Artifact) error, option *Option) error {
+	queue := list.New()
+	queue.PushBack(root)
+
+	for queue.Len() > 0 {
+		elem := queue.Front()
+		queue.Remove(elem)
+
+		artifact := elem.Value.(*Artifact)
+		if err := walkFn(artifact); err != nil {
+			return err
+		}
+
+		if !artifact.HasChildren() {
+			continue
+		}
+
+		refs, err := c.artMgr.ListReferences(ctx, q.New(q.KeyWords{"parent_id": artifact.ID}))
```

Comment:
The `Artifact` model contains the references already, no need to list them again